scylla/src/statement/prepared_statement.rs

```diff
@@ -132,59 +134,68 @@ impl PreparedStatement {
     ) -> Result<Bytes, PartitionKeyError> {
         let mut buf = BytesMut::new();
 
-        // Single-value partition key case
-        if self.get_prepared_metadata().pk_indexes.len() == 1 {
-            let pk_index = self.get_prepared_metadata().pk_indexes[0].index;
-            if let Some(v) = bound_values
-                .iter()
-                .nth(pk_index as usize)
-                .ok_or_else(|| PartitionKeyError::NoPkIndexValue(pk_index, bound_values.len()))?
-            {
-                buf.extend_from_slice(v);
-            }
-        } else {
-            // Composite partition key case
-
-            // Iterate on values using sorted pk_indexes (see deser_prepared_metadata),
-            // and use PartitionKeyIndex.sequence to insert the value in pk_values with the correct order.
-            // At the same time, compute the size of the buffer to reserve it before writing in it.
-            let mut pk_values: SmallVec<[_; 8]> =
-                smallvec![None; self.get_prepared_metadata().pk_indexes.len()];
-            let mut values_iter = bound_values.iter();
-            let mut buf_size = 0;
-            // pk_indexes contains the indexes of the partition key value, so the current offset of the
-            // iterator must be kept, in order to compute the next position of the pk in the iterator.
-            // At each iteration values_iter.nth(0) will roughly correspond to values[values_iter_offset],
-            // so values[pk_index.index] will be retrieved with values_iter.nth(pk_index.index - values_iter_offset)
-            let mut values_iter_offset = 0;
-            for pk_index in self.get_prepared_metadata().pk_indexes.iter().copied() {
-                // Find value matching current pk_index
-                let next_val = values_iter
-                    .nth((pk_index.index - values_iter_offset) as usize)
-                    .ok_or_else(|| {
-                        PartitionKeyError::NoPkIndexValue(pk_index.index, bound_values.len())
-                    })?;
-                // Add it in sequence order to pk_values
-                if let Some(v) = next_val {
-                    pk_values[pk_index.sequence as usize] = Some(v);
-                    buf_size += std::mem::size_of::<u16>() + v.len() + std::mem::size_of::<u8>();
+        let partition_key = self.extract_partition_key(bound_values)?;
+        let mut pk_val_iter = partition_key.iter().map(|(val, _spec)| val);
+        if let Some(first_value) = pk_val_iter.next() {
+            if let Some(second_value) = pk_val_iter.next() {
```

Comment:
Why do we need to do it like this? We already know the partition length, don't we? Why can't we check the length and then handle single-value or composite cases in separate branches of the `if`?

Reply:
This method is arguably more type-safe. No unwrap nor array indexing here, so no panics are possible.